core/src/main/scala/org/apache/spark/HeartbeatReceiver.scala

```diff
@@ -199,41 +222,131 @@ private[spark] class HeartbeatReceiver(sc: SparkContext, clock: Clock)
+   * function `onDisconnected`. This issue is not a `network issue` because the network is
+   * connected.
+   *
+   * [Warning 1]
+   * Worker will send heartbeats to Master every (conf.get(WORKER_TIMEOUT) * 1000 / 4) milliseconds.
+   * Check deploy/worker/Worker.scala for more details. This new mechanism design is based on the
+   * assumption: (executorTimeoutMs / 2) > (conf.get(WORKER_TIMEOUT) * 1000 / 4).
+   *
+   * [Warning 2]
+   * Not every deployment method schedules driver on master.
+   */
     logTrace("Checking for hosts with no recent heartbeats in HeartbeatReceiver.")
     val now = clock.getTimeMillis()
-    for ((executorId, lastSeenMs) <- executorLastSeen) {
-      if (now - lastSeenMs > executorTimeoutMs) {
-        logWarning(s"Removing executor $executorId with no recent heartbeats: " +
-          s"${now - lastSeenMs} ms exceeds timeout $executorTimeoutMs ms")
-        // Asynchronously kill the executor to avoid blocking the current thread
-        killExecutorThread.submit(new Runnable {
-          override def run(): Unit = Utils.tryLogNonFatalError {
-            // Note: we want to get an executor back after expiring this one,
-            // so do not simply call `sc.killExecutor` here (SPARK-8119)
-            sc.killAndReplaceExecutor(executorId)
-            // SPARK-27348: in case of the executors which are not gracefully shut down,
-            // we should remove lost executors from CoarseGrainedSchedulerBackend manually
-            // here to guarantee two things:
-            // 1) explicitly remove executor information from CoarseGrainedSchedulerBackend for
-            //    a lost executor instead of waiting for disconnect message
-            // 2) call scheduler.executorLost() underlying to fail any tasks assigned to
-            //    those executors to avoid app hang
-            sc.schedulerBackend match {
-              case backend: CoarseGrainedSchedulerBackend =>
-                backend.driverEndpoint.send(RemoveExecutor(executorId,
-                  ExecutorProcessLost(
-                    s"Executor heartbeat timed out after ${now - lastSeenMs} ms")))
-
-              // LocalSchedulerBackend is used locally and only has one single executor
-              case _: LocalSchedulerBackend =>
-
-              case other => throw new UnsupportedOperationException(
-                s"Unknown scheduler backend: ${other.getClass}")
-            }
+    if (!sc.conf.get(HEARTBEAT_RECEIVER_CHECK_WORKER_LAST_HEARTBEAT)) {
+      for ((executorId, lastSeenMs) <- executorLastSeen) {
+        if (now - lastSeenMs > executorTimeoutMs) {
+          killExecutor(executorId, now - lastSeenMs)
+          waitingList.remove(executorId)
+          executorLastSeen.remove(executorId)
+        }
+      }
+    } else {
+      for ((executorId, workerLastHeartbeat) <- waitingList) {
+        if (now - workerLastHeartbeat > executorTimeoutMs / 2) {
```

Comment:
why `executorTimeoutMs / 2` ?

Reply:
If an executor lasts in `waitingList` for more than `executorTimeoutMs / 2` seconds, it will be killed. I just want to choose a value less than `executorTimeoutMs`, and thus I choose `executorTimeoutMs / 2`. Whether we need an additional variable to set the value of timeout in `waitingList`?

Reply:
Ideally, yes - since they mean different things; though I can see how closely related they are. +CC @Ngone51 thoughts ?

At a minimum, we should document what the config means, and what implications are of setting it in the config's doc.

Reply:
Add a new config in https://github.com/apache/spark/pull/37411/commits/0c409344877efb481d91d5190f4d801a39910a72.